### ISSUE TITLE
Implementing simple neural attention model (for padded inputs)

### ISSUE THREAD

**Issue by anayebi** [fchollet/keras#2612]
I have an input sequence (because it is variable length, used the Keras `pad_sequences` function to zero pad it), and I wanted to basically average the outputs of my RNN and concatenate it with its prediction at the last timestep, which would then be the model's prediction.

Here is my code to better visualize what I mean (I'm using the Keras 1.0 Model class):

>   input_sentences = Input(shape=(max_sen_length,), dtype='int32')
     x = Embedding(input_dim=vocab_size, output_dim=vocab_dim, input_length=max_sen_length, mask_zero=True)(input_sentences)
    lstm_out = LSTM(150, return_sequences=True)(x)
    attention = TimeDistributedDense(32, activation='softmax')(lstm_out)
    context = Lambda(avg_timesteps, output_shape=avg_timesteps_output_shape)(attention)
    last_lstm_out = Lambda(last_timestep, output_shape=last_timestep_output_shape)(lstm_out)
    concat = Merge([last_lstm_out, context], mode='concat')

Note that `avg_timesteps` and `last_timestep` were custom functions I wrote to work with the `Lambda` layers (I won't include them here for brevity).

The issue I am running into is that `Lambda` layers do **not** support masking, so I was not able to get the above code to work. Is there a way to essentially implement my context layer (which is the average of the outputs of the `TimeDistributedDense` layer)? 

I am aware of the various Keras threads about implementing neural attention, but this is a relatively simple version of that (since my model's overall prediction is based on the last timestep, so I'm just averaging the predictions up to then and concatenating them with the RNN's last output, `last_lstm_out`), based on Section 3.1 of [this paper](http://arxiv.org/pdf/1508.04025v5.pdf). I figure this is probably doable, so I'd be happy to hear anyone's thoughts on this. Thanks!

**Comment by braingineer** [comment]
I have one here that supports masking:

https://gist.github.com/braingineer/27c6f26755794f6544d83dec2dd27bbb

Note that after attention, there's no need for a mask (it's collapsed across the masked dimension)

**Comment by braingineer** [comment]
if you need access to both the probability distribution AND the attention, you can do the ProbabilityTensor in that gist and then use it with 

```python
class EZAttend(Layer):
    def __init__(self, p_tensor, *args, **kwargs):
        self.supports_masking = True
        self.p_tensor = p_tensor
        super(EZAttend, self).__init__(*args, **kwargs)

    def compute_mask(self, x, mask=None):
        return None

    def get_output_shape_for(self, input_shape):
        last_dim = K.ndim(self.p_tensor)
        output_shape = list(input_shape)
        output_shape.pop(last_dim-1)
        return tuple(output_shape)

    def call(self, target_tensor, mask=None):
        last_dim = K.ndim(self.p_tensor)
        expanded_p = K.repeat_elements(K.expand_dims(self.p_tensor, last_dim), 
                                       K.shape(target_tensor)[last_dim], 
                                       axis=last_dim)
        return K.sum(expanded_p * target_tensor, axis=last_dim-1)
```

**Comment by braingineer** [comment]
also, I have a lambda layer with masking that I haven't had time to do a PR for (I put masking on a lot of the core.py layers because I do a lot of masked sequence stuff)

https://gist.github.com/braingineer/b64ca35223c7782667984d34ddb7a7fa

**Comment by anayebi** [comment]
Thanks, @braingineer, I will check these out!

**Comment by anayebi** [comment]
@braingineer thanks again, your implementation was very helpful, and I was able to use it (citing your code, of course) in a course project to implement a new recurrent attention model, found here: https://github.com/anayebi/CS-224U-Project

Specifically, the file `softattention.py` uses the code you posted and adds on additional custom layers for neural attention (specifically recurrent/LSTM-like attentional layers).
